### PR TITLE
252 recurring booking

### LINKED ISSUE #252: Bookings can be recurring and out of series
### User story
Bookings can be recurring and out of series
### Narrative
As a user when I create a booking request I have the option to make it reccurring daily, weekly or monthly, while respecting the maximum amount of bookings allowed for the room/resource in question
### Acceptance criteria

| # | Given | When | Then |
|---|-------|------|------|
|1|I am an authenticated user|I book a room recurringly while respecting my bookings amounts|the room is booked recurringly |
|2|I am an authenticated user|I book a room recurringly while going above my allowed booking amounts|the booking requested is denied |
|3|I am an authenticated user|I book a room out of series while respecting my bookings amounts|the room is booked |
|4|I am an authenticated user|I book a room out of series while going above my allowed booking amounts|the booking requested is denied |

### PR DESCRIPTION
**Pull request by EvanDime**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**
Please provide enough information so that others can review your pull request:
Complete rework of booking requests to now use reservation as a pivot table for rooms. 

Explain the **details** for making this change. What existing problem does the pull request solve?
Allows for multiple reservations to be part of a booking request

can add and remove dates from a creation. 
### Test plan (required)
Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI.

Screen Shot 2021-01-25 at 4 33 39 AM
Screen Shot 2021-01-25 at 4 27 13 AM

### Closing issues
List all issues this PR resolves.
- closes #252 

### Pre-Merge checklist 
Complete these items before requesting reviews
- [x] Static Analysis Rules Validated
- [x] All commits follow naming conventions 

### Sub-tasks Checklist
- [x] UI Implementation
- [x] UI Test
- [x] Backend Implementation
- [x] Backend Tests
- [x] Integration Tests
- [x] System Tests
- [x] Documentation

### CODE REVIEW

**Comment by EvanDime** [comment]
Currently needs some documentation updates but feel free to review.

**Comment by EvanDime** [comment]
I'll start adding checks for conflicts to this PR. I also need to make sure the maximum is not set.

**Review by Alex-Hayeur — APPROVED** [review]
Mmm that sweet 100% code coverage.
Looks good, and a fantastic rework of how our bookings are handled. This for sure will require some documentation changes though.
Just to double-check what I've seen, what this is doing is basically turning booking requests into a relational table between the reservations and rooms, and now most of the old information is in reservations + the new recurrences functionality?

File: database/seeders/ReservationsSeeder.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+class ReservationsSeeder extends Seeder
+{
+    /**
+     * Run the database seeds.
+     *
+     * @return void
+     */
+    public function run()
+    {
+
```
Comment (WindOfRussia) [inline database/seeders/ReservationsSeeder.php:17]: should we not generate some right away?

File: database/seeders/DatabaseSeeder.php [hunk 2]
```diff
@@ -21,13 +21,14 @@ public function run()
             $this->call([
                 EasyUserSeeder::class
             ]);
+            User::first()->assignRole('super-admin');
```
Comment (WindOfRussia) [inline database/seeders/DatabaseSeeder.php:24]: nice

File: database/factories/ReservationFactory.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+    /**
+     * Define the model's default state.
+     *
+     * @return array
+     */
+    public function definition()
+    {
+        return [
+            //
```
Comment (WindOfRussia) [inline database/factories/ReservationFactory.php:25]: should probably copy the definitions from the BookingRequest factory

File: app/Http/Controllers/BookingRequestController.php [hunk 4]
```diff
@@ -167,6 +177,7 @@ public function update(Request $request, BookingRequest $booking)
      */
     public function destroy(BookingRequest $booking)
     {
+        Reservation::where('booking_request_id', $booking->id)->delete();
```
Comment (WindOfRussia) [inline app/Http/Controllers/BookingRequestController.php:180]: maybe observer?

File: app/Models/Reservation.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+
+class Reservation extends Model
```
Comment (WindOfRussia) [inline app/Models/Reservation.php:8]: extend Pivot since this the in-between table from bookings and rooms

File: app/Models/Room.php [hunk 6]
```diff
@@ -133,5 +121,64 @@ public function verifyDatesAreWithinRoomRestrictions($startDate)
...
+      $query = $query->where('id','!=', $reservation->id);
+    }
+    $query = $query->where(function ($query) use ($startDate, $endDate) {
+      $query->whereBetween('start_time', [$startDate, $endDate])
+      ->orWhereBetween('end_time', [$startDate, $endDate])
+      ->orWhere(function ($query) use ($startDate, $endDate) {
+        $query->where('start_time', '<', $startDate)->where('end_time', '>', $endDate);
+      });
+    });
+      $conflict = $query->first();
```
Comment (WindOfRussia) [inline app/Models/Room.php:157]: count() query check could be faster here instead of loading the real model

**Review by WindOfRussia — APPROVED** [review]
thats a big yes from me..

**Review by jonathanp16 — APPROVED** [review]
Great job! Very clean refactoring of the booking request model

**Pull request merged**